Initial commit of the TDM Reader swagger API

BKnight760 opened the pull request:
Initial commit of the nitdmreader.yml swagger doc.  This is used by the TDM Reader service, but is not code-generated into a a webservice interface yet.

[x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/systemlink-OpenAPI-documents/blob/master/CONTRIBUTING.md).

### What does this Pull Request accomplish?

This pull request adds a Swagger document for the TDM Reader service. This will be the master copy of the TDM Reader service's Swagger document, for use in SystemLink Server.

### Why should this Pull Request be merged?

This pull request gives higher visibility to the TDM Reader HTTP API.

### What testing has been done?

Existing automated tests have been run against webservice, and all tests pass.

BKnight760 committed:
Initial commit of the nitdmreader.yml swagger doc

Initial commit of the nitdmreader.yml swagger doc.  This is used by the TDM Reader service, but is not code-generated into a a webservice interface yet.

spanglerco commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,508 @@
+        type: integer
+        description: Count of all properties. This could be more than the items returned in Properties.
+        example: 1
+  FileList:
```
This doesn't have the latest change made last Thursday that wraps this in an object.

adelcast commented:
Should we change the section name for Properties to something more meaningful like Metadata? The problem is that Properties already have a meaning (Channels, Files and Groups have properties). The query-files endpoint returns metadata as well as properties (maybe the endpoint name should change too?)

BKnight760 committed:
Merge previous changes.

Merged back changes for earlier commits in VSTS.

BKnight760 committed:
Rename 'properties' to 'metadata'

Rename 'properties' to 'metadata' in the tags element for file, group, and channel.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+    type: object
+    properties:
+      fileIds:
+        description: File Ids
```
IDs

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+    properties:
+      decimatex:
+        type: integer
+        description: Number of pixels on the x axis which is the number of points that can be displayed.
```
Number of pixels on the x axis. This determines the number of points that can be displayed.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        format: int32
+      decimatey:
+        type: integer
+        description: Number of pixels on the y axis which is the number of points that can be displayed
```
Number of pixels on the y axis. This determines the number of points that can be displayed.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        type: number
+        default: 0
+        format: double
+        description: Minimum X value of the display X axis.
```
Do we want to capitalize X/Y or no? Either is fine, just need to make sure they're consistent. They're capitalized here, but not above.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+      showPoints:
+        type: boolean
+        default: false
+        description: Decimate to point style graph.
```
Here and in the previous description, I'm not sure even after some googling what decimate means in this context. Is there a way we could put this in plainer terms for newer users? "Represent data as a point-style graph" or something?

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+      - groupName
+      - channelName
+  ChannelSpecifications:
+    description: Specification for an array of groups of channels potentially with x-y relationships.
```
Could "groups of channels" be "channel groups?" That'd get rid of one of the prepositional phrases and make this a little easier to follow.

Also curious what "potentially" means here, so there's an opportunity for more clarity. Does it mean the groups of channels support these relationships?

BKnight760 replied:
The wording was confusing.  The object here is an array of x-y channel groupings.  Each grouping contains 0 to 1 x-channels, and N y-channels.  The x-channel is optional, which is what the "potentially" was referring to.  Hopefully the new wording is clearer.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+          required:
+            - y
+  ChannelSpecificationsAndWindow:
+    description: Channel specifications and data window.
```
Can remove period

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+    properties:
+      dataType:
+        type: string
+        description: Native data type of data
```
Maybe "Native data type for the channel" or "Native data type of data in the channel"

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+          type: string
+      decimated:
+        type: boolean
+        description: is returned data decimated
```
Specifies whether returned data is decimated [though again, there might be a clearer way to describe decimation]

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        description: Native data type of data
+      numericData:
+        type: array
+        description: channel data
```
Capital C to match capitalization in other descriptions. This is one of those things that doesn't _really_ matter, but consistency is a good way to build trust with the power Swagger user type.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        description: is returned data decimated
+      undecimatedLength:
+        type: integer
+        description: number of points of undecimated data
```
Number

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        $ref: '#/definitions/OneChannelData'
+      y:
+        type: array
+        description: y channels
```
Depending on whether you decide to capitalize x and y throughout, might want to capitalize Y here.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+            $ref: '#/definitions/XYData'
+      timeUtcOffsetHours:
+        type: number
+        description: The timezone offset applied to local time values
```
time zone

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        type: number
+        description: The timezone offset applied to local time values
+  NamesList:
+    description: A list of names and a count of all names.
```
"List.." is fine

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+          example: Channel 1
+      totalCount:
+        type: integer
+        description: Count of all items in the list. This could be more than the items returned in Names.
```
Total number of items in the list. This might be more than the number names.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+          type: integer
+          default: -1
+        - name: returnCount
+          description: Specify this to return the total count of items
```
Specifies whether to return the total number of items

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+  /v1/files/{file-id}/groups/{group-name}/channels:
+    get:
+      tags: [metadata]
+      summary: Lists all channels in the group
```
Other Swagger documents I've reviewed have both a summary and description. Is there a specific reason to only have summary? I don't know a ton about implementation. Could maybe ask Phil about this.

jjaco2 commented on tdm-reader/nitdmreader.yml:
```diff
@@ -0,0 +1,512 @@
+        - application/json
+      parameters:
+        - name: ChannelSpecificationsAndWindow
+          description: Channel specifications and window information.
```
Do we expect users to know what window information is? Just something to think about

jjaco2 reviewed (commented):
Just some questions to clarify wording and some notes for consistency.

In addition to the things I noted, consider removing periods from any descriptions that aren't a complete sentence. This seems to be the industry-standard way to punctuate in these documents.

BKnight760 committed:
Update descriptions per review feedback

BKnight760 committed:
Addressed more review comments

Clarified descriptions from review comments
